### PR TITLE
Fix #2894 - Profiler Processor & Metrics

### LINKED ISSUE #2894: Profiler processor and status
**Feature**
Based on https://github.com/open-metadata/OpenMetadata/issues/1311

**Describe the task**
Prepare the proper Processor and Status for the ORM profiler

### PR DESCRIPTION
**Pull request by pmbrull**
### Describe your changes :

This PR fixes https://github.com/open-metadata/OpenMetadata/issues/2894 and partially https://github.com/open-metadata/OpenMetadata/issues/2878

There are two main changes in this PR:
- Instead of relying on all the profiling and validation logic in the Workflow, we have prepared a Processor to handle that stuff, trying to follow a more strict `workflow` logic. We have also updated the Status to properly log the results.
- We have implement new metrics:
    - AVG - which also handles AVG_LENGTH for strings and text
    - DISTINCT_COUNT
    - DUPLICATE_COUNT
    - HISTOGRAM -> This is interesting because it helped us notice that we were missing a more complex type of Metric `QueryMetric`.  We have also developed intermediate functions (CONCAT) that can be used when computing other metrics. 

#
### Type of change :

- [ ] Bug fix
- [x] Improvement
- [x] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Frontend Preview (Screenshots) :
<p align="center">For frontend related change, please link screenshots of your changes preview! Optional for backend related changes.
</p>

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/open-source-community/developer) document.
- [ ] I have performed a self-review of my own. 
- [ ] I have tagged my reviewers below.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#

### CODE REVIEW

File: ingestion/src/metadata/ingestion/api/processor.py [hunk 1]
```diff
@@ -41,7 +41,11 @@ class Processor(Closeable, Generic[Entity], metaclass=ABCMeta):
     @classmethod
     @abstractmethod
     def create(
-        cls, config_dict: dict, metadata_config_dict: dict, ctx: WorkflowContext
+        cls,
+        config_dict: dict,
+        metadata_config_dict: dict,
+        ctx: WorkflowContext,
+        **kwargs
```
Comment (pmbrull) [inline ingestion/src/metadata/ingestion/api/processor.py:48]: Allows us to pass more parameters to the processor, such as the `session` in the ORM Processor

File: ingestion/tests/unit/profiler/test_metrics.py [hunk 2]
```diff
@@ -111,3 +125,72 @@ def test_table_count(self):
...
+            num = Column(Integer)
+
+        TestHist.__table__.create(bind=self.engine)
+
+        data = [TestHist(num=int(rand)) for rand in normal(loc=0, scale=10, size=2000)]
+
+        self.session.add_all(data)
+        self.session.commit()
+
+        hist = Metrics.HISTOGRAM(TestHist.num, bins=5)
```
Comment (pmbrull) [inline ingestion/tests/unit/profiler/test_metrics.py:192]: We can now pass kwargs when creating metrics to be more flexible in internal functionalities. Still needs to be updated in the Processor

File: ingestion/src/metadata/orm_profiler/orm/registry.py [hunk 3]
```diff
@@ -25,23 +27,42 @@ class CustomTypes(TypeRegistry):
...
+
+
+def is_numeric(_type) -> bool:
+    """
+    Check if sqlalchemy _type is derived from Numeric
+    """
+    return issubclass(_type.__class__, Numeric)
+
+
+def is_quantifiable(_type) -> bool:
```
Comment (pmbrull) [inline ingestion/src/metadata/orm_profiler/orm/registry.py:56]: helps handle what types we support for each metric

**Comment by pmbrull** [comment]
> [**open-metadata-ingestion**] Kudos, SonarCloud Quality Gate passed!    Quality Gate passed
> 
> [Bug](https://sonarcloud.io/project/issues?id=open-metadata-ingestion&pullRequest=2900&resolved=false&types=BUG) [C](https://sonarcloud.io/project/issues?id=open-metadata-ingestion&pullRequest=2900&resolved=false&types=BUG) [1 Bug](https://sonarcloud.io/project/issues?id=open-metadata-ingestion&pullRequest=2900&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=open-metadata-ingestion&pullRequest=2900&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=open-metadata-ingestion&pullRequest=2900&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=open-metadata-ingestion&pullRequest=2900&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=open-metadata-ingestion&pullRequest=2900&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=open-metadata-ingestion&pullRequest=2900&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=open-metadata-ingestion&pullRequest=2900&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=open-metadata-ingestion&pullRequest=2900&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=open-metadata-ingestion&pullRequest=2900&resolved=false&types=CODE_SMELL) [0 Code Smells](https://sonarcloud.io/project/issues?id=open-metadata-ingestion&pullRequest=2900&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=open-metadata-ingestion&pullRequest=2900&metric=new_coverage&view=list) [0.0% Coverage](https://sonarcloud.io/component_measures?id=open-metadata-ingestion&pullRequest=2900&metric=new_coverage&view=list) [](https://sonarcloud.io/component_measures?id=open-metadata-ingestion&pullRequest=2900&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=open-metadata-ingestion&pullRequest=2900&metric=new_duplicated_lines_density&view=list)

Sonar is complaining about the join condition `and_(1 == 1)`. I could not figure out any other way of replicating this

File: ingestion/src/metadata/config/workflow.py [hunk 4]
```diff
@@ -114,3 +115,43 @@ def get_sink(
...
+    the source tables.
+
+    :param processor_type: Type specified in the config, e.g., metadata-rest
+    :param context: Workflow related information
+    :param processor_config: Specific Processor configurations, such as the profiler and tests
+    :param metadata_config: Metadata server configurations
+    :param _from: From where do we load the sink class. Ingestion by default.
+    """
+    processor_class = get_class(
+        "metadata.{}.processor.{}.{}Processor".format(
```
Comment (harshach) [inline ingestion/src/metadata/config/workflow.py:144]: We need to make this more flexible to load the class without depending on the packaging convention. 
Not important right now but can be annoying for users to write their own sources, processors

**Review by harshach — APPROVED** [review]

File: ingestion/src/metadata/config/workflow.py (hunk 4 above)
Comment (pmbrull) [reply ingestion/src/metadata/config/workflow.py:144]: That's a good point. Let's tackle this together with https://github.com/open-metadata/OpenMetadata/issues/2848. I'll add a comment there

**Pull request merged**